nova/scheduler/weights/affinity.py

```diff
@@ -70,6 +77,36 @@ def weight_multiplier(self):
 
         return CONF.filter_scheduler.soft_affinity_weight_multiplier
 
+    def _weigh_object(self, host_state, request_spec):
+        weight = super(ServerGroupSoftAffinityWeigher, self)._weigh_object(
+            host_state, request_spec)
+
+        # if the host contained servers from the same group, return that weight
+        if weight:
+            return weight
+
+        # get shards of the host
+        host_shard_aggr = [aggr for aggr in host_state.aggregates
+                            if aggr.name.startswith(self._SHARD_PREFIX)]
+        if len(host_shard_aggr) > 1:
+            LOG.warning('More than one host aggregates found for '
+                        'host %(host)s, selecting first.',
+                        {'host': host_state.host})
+            host_shard_aggr = [host_shard_aggr[0], ]
+
+        group_hosts = None
+        if request_spec.instance_group and request_spec.instance_group.hosts:
+            group_hosts = set(request_spec.instance_group.hosts)
+        if not group_hosts:
+            return 0
+        # if any host in the server-group's hosts is part of this aggregate
+        for host in group_hosts:
+            if self._is_host_in_aggregate(host, host_shard_aggr):
+                weight = 0.5
+                break
```

Comment:
```python
# group_hosts doesn't contain our host, because otherwise we would have returned a weighed already as we would have instances
if group_hosts & set(host_shard_aggr.hosts):
    return 0.5
return 0
```
This is what I meant with set operations. Then we don't need to for-loop over the group_hosts and also not over the aggregate.hosts all the time.

Reply:
ok fixed and removed _is_host_in_aggregate().